File: src/main/java/com/leet/medium/AllPathsFromSourceToTarget.exs
defmodule Solution do
  def solve_node(adj, ans, path, node, dst) do
    if node == dst do
      [Enum.reverse([node | path]) | ans]
    else
      Map.get(adj, node)
      |> Enum.reduce(ans, fn sub, acc ->
        solve_node(adj, acc, [node | path], sub, dst)
      end)
    end
  end

  def solve_node_dp(adj, node, dp, dst) do
    cond do
      node == dst ->
        [[node]]

      Map.has_key?(dp, node) ->
        Map.get(dp)

      true ->
        res =
          Map.get(adj, node)
          |> Enum.reduce([], fn sub, acc ->
            solve_node(adj, sub, dp, dst)
            |> Enum.reduce(acc, fn lis, coll -> [[node | lis] | coll] end)
          end)

        Map.put(dp, node, res)
        res
    end
  end

  @spec all_paths_source_target(graph :: [[integer]]) :: [[integer]]
  def all_paths_source_target(graph) do
    {adj, size} =
      Enum.reduce(
        graph,
        {Map.new(), 0},
        fn sub, {adj, idx} -> {Map.put(adj, idx, sub), idx + 1} end
      )

    solve_node(adj, [], [], 0, size - 1)
  end
end

File: src/main/java/com/leet/medium/AllPathsFromSourceToTarget.rs
struct Solution {}

impl Solution {
  fn dfs(
    graph: &Vec<Vec<i32>>,
    node: i32,
    idx: usize,
    path: &mut Vec<i32>,
    ans: &mut Vec<Vec<i32>>,
  ) {
    if idx >= path.len() {
      path.push(node);
    } else {
      path[idx] = node;
    }

    if node == graph.len() as i32 - 1 {
      // println!("ans = {:?}", &path[0..=idx]);
      ans.push(path[0..=idx].to_vec().clone());
    } else {
      for &next in graph[node as usize].iter() {
        Self::dfs(graph, next, idx + 1, path, ans);
      }
    }
  }

  pub fn all_paths_source_target(graph: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let mut ans = vec![];
    Self::dfs(&graph, 0, 0, &mut vec![], &mut ans);
    ans
  }
}

fn main() {
  let graph = vec![vec![4, 3, 1], vec![3, 2, 4], vec![3], vec![4], vec![]];
  let ans = Solution::all_paths_source_target(graph);
  println!("{:?}", ans);
}
